src/dragonball/src/device_manager/vhost_net_dev_mgr.rs

```diff
@@ -0,0 +1,646 @@
+        if config.num_queues % 2 != 0 {
+            return Err(VhostNetDeviceError::InvalidQueueNum(config.num_queues));
+        }
+        if !cfg!(feature = "hotplug") && ctx.is_hotplug {
+            return Err(VhostNetDeviceError::UpdateNotAllowedPostBoot);
+        }
+
+        slog::info!(
+            ctx.logger(),
+            "add vhost-net device configuration";
+            "subsystem" => "vhost_net_dev_mgr",
+            "id" => &config.iface_id,
+            "host_dev_name" => &config.host_dev_name,
+        );
+
+        let mgr = &mut device_mgr.vhost_net_manager;
+        let device_index = mgr.info_list.insert_or_update(&config)?;
+
+        // If it is a hotplug device, then it will be attached immediately.
+        if ctx.is_hotplug {
+            slog::info!(
+                ctx.logger(),
+                "attach vhost-net device";
+                "subsystem" => "vhost_net_dev_mgr",
+                "id" => &config.iface_id,
+                "host_dev_name" => &config.host_dev_name,
+            );
+
+            match Self::create_device(&config, &mut ctx) {
+                Ok(device) => {
+                    let mmio_dev = DeviceManager::create_mmio_virtio_device(
+                        device,
+                        &mut ctx,
+                        config.use_shared_irq.unwrap_or(mgr.use_shared_irq),
+                        config.use_generic_irq.unwrap_or(USE_GENERIC_IRQ),
+                    )
+                    .map_err(VhostNetDeviceError::RegisterNetDevice)?;
+                    ctx.insert_hotplug_mmio_device(&mmio_dev, None)
+                        .map_err(VhostNetDeviceError::DeviceManager)?;
+                    // live-upgrade need save/restore device from info.device.
+                    mgr.info_list[device_index].set_device(mmio_dev);
+                }
+                Err(err) => {
+                    mgr.info_list.remove(device_index);
+                    return Err(VhostNetDeviceError::Virtio(err));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Attach all configured vhost-net device to the virtual machine instance.
+    pub fn attach_devices(&mut self, ctx: &mut DeviceOpContext) -> Result<(), VhostNetDeviceError> {
+        for info in self.info_list.iter_mut() {
+            slog::info!(
+                ctx.logger(),
+                "attach vhost-net device";
+                "subsystem" => "vhost_net_dev_mgr",
+                "id" => &info.config.iface_id,
+                "host_dev_name" => &info.config.host_dev_name,
+            );
+
+            let device = Self::create_device(&info.config, ctx)
+                .map_err(VhostNetDeviceError::CreateNetDevice)?;
+            let mmio_dev = DeviceManager::create_mmio_virtio_device(
+                device,
+                ctx,
+                info.config.use_shared_irq.unwrap_or(self.use_shared_irq),
+                info.config.use_generic_irq.unwrap_or(USE_GENERIC_IRQ),
```

Comment:
we don't have generic irq now, should we add this?

Reply:
I'll remove all generic irq from Dragonball in the later PR.

Reply:
see #8342